src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -1164,6 +996,377 @@ void DiscoveryDataBase::AckedFunctor::operator () (
+                reader_it != tit->second.end();
+                ++reader_it)
+        {
+            if (*reader_it == reader_guid)
+            {
+                tit->second.erase(reader_it);
+                break;
+            }
+        }
+
+        if (tit->second.empty())
+        {
+            readers_by_topic_.erase(tit);
+        }
+    }
+}
+
+void DiscoveryDataBase::add_writer_to_topic_(const eprosima::fastrtps::rtps::GUID_t& writer_guid,
+        const std::string& topic_name)
+{
+    // Update writers by topic
+    std::map<std::string, std::vector<eprosima::fastrtps::rtps::GUID_t> >::iterator topic_it =
+            writers_by_topic_.find(topic_name);
+    if (topic_it != writers_by_topic_.end())
+    {
+        std::vector<eprosima::fastrtps::rtps::GUID_t>::iterator writer_by_topic_it =
+                std::find(topic_it->second.begin(), topic_it->second.end(), writer_guid);
+        if (writer_by_topic_it == topic_it->second.end())
+        {
+            logInfo(DISCOVERY_DATABASE, "New writer " << writer_guid << " in writers_by_topic: " << topic_name);
+            topic_it->second.push_back(writer_guid);
+        }
+        else
+        {
+            *writer_by_topic_it = writer_guid;
+        }
+    }
+    // This is the first writer in the topic
+    else
+    {
+        std::vector<fastrtps::rtps::GUID_t> writers_in_topic = {writer_guid};
+        auto retu = writers_by_topic_.insert(
+            std::pair<std::string, std::vector<fastrtps::rtps::GUID_t>>(topic_name, writers_in_topic));
+        if (!retu.second)
+        {
+            logError(DISCOVERY_DATABASE, "Could not insert writer " << writer_guid << " in topic " << topic_name);
+        }
+        else
+        {
+            logInfo(DISCOVERY_DATABASE, "New topic in readers_by_topic: " << topic_name);
+        }
+
+    }
+}
+
+void DiscoveryDataBase::add_reader_to_topic_(const eprosima::fastrtps::rtps::GUID_t& reader_guid,
+        const std::string& topic_name)
+{
+    // Update readers_by_topic
+    std::map<std::string, std::vector<eprosima::fastrtps::rtps::GUID_t> >::iterator topic_it =
+            readers_by_topic_.find(topic_name);
+    if (topic_it != readers_by_topic_.end())
+    {
+        std::vector<eprosima::fastrtps::rtps::GUID_t>::iterator reader_by_topic_it =
+                std::find(topic_it->second.begin(), topic_it->second.end(), reader_guid);
+        if (reader_by_topic_it == topic_it->second.end())
+        {
+            logInfo(DISCOVERY_DATABASE, "New reader " << reader_guid << " in readers_by_topic: " << topic_name);
+            topic_it->second.push_back(reader_guid);
+        }
+        else
+        {
+            *reader_by_topic_it = reader_guid;
+        }
+    }
+    // This is the first reader in the topic
+    else
+    {
+        std::vector<fastrtps::rtps::GUID_t> readers_in_topic = {reader_guid};
+        auto retu = readers_by_topic_.insert(
```

Comment:
why not?
`readers_by_topic_[topic_name] = vector<GUID_t>{reader_guid};`

Reply:
:+1: